`Error: No viable remote package found` fallback wanted: `cargo install`

**TL;DR, `binstall` fails when `quickinstall` release is missing or when other errors occur. Instead of failing, I'd like a fallback option that tries `cargo install` before giving up.** 

Overall this project is really promising! I have a dozen or so changes like this that I want to make to my Dockerfile:

```diff
-wget -q https://github.com/watchexec/watchexec/releases/download/cli-v1.18.6/watchexec-1.18.6-x86_64-unknown-linux-musl.tar.xz -O - | tar xJ watchexec-1.18.6-x86_64-unknown-linux-musl/watchexec --strip-components=1
+cargo binstall --no-confirm watchexec --version 1.18.6
```

However, I've found that several of my packages neither have `binstall` metadata nor are they cached on `https://github.com/alsuren/cargo-quickinstall/releases`.

```
cargo binstall --no-confirm watchexec                 
14:57:13 [INFO] Installing package: 'watchexec'
14:57:13 [INFO] Checking for package at: 'https://github.com/watchexec/watchexec/releases/download/v1.17.1/watchexec-x86_64-unknown-linux-gnu-v1.17.1.tgz'
14:57:14 [INFO] Sending installation report to quickinstall (anonymous)
14:57:15 [INFO] Checking for package at: 'https://github.com/alsuren/cargo-quickinstall/releases/download/watchexec-1.17.1-x86_64-unknown-linux-gnu/watchexec-1.17.1-x86_64-unknown-linux-gnu.tar.gz'
14:57:16 [ERROR] File does not exist remotely, cannot proceed
Error: No viable remote package found
```


For context, here's the [cargo-quickinstall docs](https://github.com/alsuren/cargo-quickinstall/tree/459b2b5d66e58f8cfe9c36d652c3a95fea813fe9#usage):

> If we don't have a pre-compiled version, it will fallback to `cargo install` automatically.

Does `binstall` intentionally lack that behavior? If so, it might be nice to have a note about that on the README; I expected a fallback, considering this project's relationship to `quickinstall`. If not, consider this a feature request.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,29 +19,29 @@ pkg-fmt = "zip"
 pkg-fmt = "zip"
 
 [dependencies]
-crates_io_api = "0.8.0"
-cargo_metadata = "0.14.1"
-tinytemplate = "1.2.1"
-tokio = { version = "1.16.1", features = [ "full" ] }
-log = "0.4.14"
-structopt = "0.3.26"
-simplelog = "0.11.2"
-anyhow = "1.0.53"
-reqwest = { version = "0.11.9", features = [ "rustls-tls" ], default-features = false }
-tempdir = "0.3.7"
-flate2 = "1.0.22"
-tar = "0.4.38"
-cargo_toml = "0.11.4"
-serde = { version = "1.0.136", features = [ "derive" ] }
-strum_macros = "0.23.1"
-strum = "0.23.0"
-dirs = "4.0.0"
-crates-index = "0.18.5"
-semver = "1.0.7"
-xz2 = "0.1.6"
-zip = "0.5.13"
+anyhow = "1.0.57"
 async-trait = "0.1.52"
+cargo_metadata = "0.14.2"
+cargo_toml = "0.11.4"
+crates-index = "0.18.7"
+crates_io_api = "0.8.0"
+dirs = "4.0.0"
+flate2 = "1.0.22"
+log = "0.4.14"
+reqwest = { version = "0.11.10", features = [ "rustls-tls" ], default-features = false }
+semver = "1.0.7"
+serde = { version = "1.0.136", features = [ "derive" ] }
+simplelog = "0.12.0"
+structopt = "0.3.26"
+strum = "0.24.0"
+strum_macros = "0.24.0"
+tar = "0.4.38"
+tempdir = "0.3.7"
+tinytemplate = "1.2.1"
+tokio = { version = "1.18.0", features = [ "full" ] }
 url = "2.2.2"
+xz2 = "0.1.6"
+zip = "0.6.2"
 
 [dev-dependencies]
 env_logger = "0.9.0"

--- a/src/main.rs
+++ b/src/main.rs
@@ -1,11 +1,11 @@
 use std::{path::PathBuf, str::FromStr};
 
+use cargo_toml::{Package, Product};
 use log::{debug, error, info, warn, LevelFilter};
 use simplelog::{ColorChoice, ConfigBuilder, TermLogger, TerminalMode};
-
 use structopt::StructOpt;
-
 use tempdir::TempDir;
+use tokio::process::Command;
 
 use cargo_binstall::{
     bins,
@@ -121,7 +121,8 @@ async fn main() -> Result<(), anyhow::Error> {
     let (mut meta, binaries) = (
         package
             .metadata
-            .map(|m| m.binstall)
+            .as_ref()
+            .map(|m| m.binstall.clone())
             .flatten()
             .unwrap_or(PkgMeta::default()),
         manifest.bin,
@@ -160,11 +161,34 @@ async fn main() -> Result<(), anyhow::Error> {
     fetchers.add(GhCrateMeta::new(&fetcher_data).await);
     fetchers.add(QuickInstall::new(&fetcher_data).await);
 
-    let fetcher = fetchers.first_available().await.ok_or_else(|| {
-        error!("File does not exist remotely, cannot proceed");
-        anyhow::anyhow!("No viable remote package found")
-    })?;
+    match fetchers.first_available().await {
+        Some(fetcher) => {
+            install_from_package(
+                binaries,
+                fetcher,
+                install_path,
+                meta,
+                opts,
+                package,
+                pkg_path,
+                temp_dir,
+            )
+            .await
+        }
+        None => install_from_source(opts, package).await,
+    }
+}
 
+async fn install_from_package(
+    binaries: Vec<Product>,
+    fetcher: &dyn Fetcher,
+    install_path: PathBuf,
+    mut meta: PkgMeta,
+    opts: Options,
+    package: Package<Meta>,
+    pkg_path: PathBuf,
+    temp_dir: TempDir,
+) -> Result<(), anyhow::Error> {
     // Prompt user for third-party source
     if fetcher.is_third_party() {
         warn!(
@@ -292,4 +316,42 @@ async fn main() -> Result<(), anyhow::Error> {
 
     info!("Installation complete!");
     Ok(())
+}
+
+async fn install_from_source(opts: Options, package: Package<Meta>) -> Result<(), anyhow::Error> {
+    // Prompt user for source install
+    warn!("The package will be installed from source (with cargo)",);
+    if !opts.no_confirm && !opts.dry_run && !confirm()? {
+        warn!("Installation cancelled");
+        return Ok(());
+    }
+
+    if opts.dry_run {
+        info!(
+            "Dry-run: running `cargo install {} --version {}`",
+            package.name, package.version
+        );
+        Ok(())
+    } else {
+        debug!(
+            "Running `cargo install {} --version {}`",
+            package.name, package.version
+        );
+        let mut child = Command::new("cargo")
+            .arg("install")
+            .arg(package.name)
+            .arg("--version")
+            .arg(package.version)
+            .spawn()?;
+        debug!("Spawned command pid={:?}", child.id());
+
+        let status = child.wait().await?;
+        if status.success() {
+            info!("Cargo finished successfully");
+            Ok(())
+        } else {
+            error!("Cargo errored! {:?}", status);
+            Err(anyhow::anyhow!("Cargo install error"))
+        }
+    }
 }

--- a/src/main.rs
+++ b/src/main.rs
@@ -328,20 +328,22 @@ async fn install_from_source(opts: Options, package: Package<Meta>) -> Result<()
 
     if opts.dry_run {
         info!(
-            "Dry-run: running `cargo install {} --version {}`",
-            package.name, package.version
+            "Dry-run: running `cargo install {} --version {} --target {}`",
+            package.name, package.version, opts.target
         );
         Ok(())
     } else {
         debug!(
-            "Running `cargo install {} --version {}`",
-            package.name, package.version
+            "Running `cargo install {} --version {} --target {}`",
+            package.name, package.version, opts.target
         );
         let mut child = Command::new("cargo")
             .arg("install")
             .arg(package.name)
             .arg("--version")
             .arg(package.version)
+            .arg("--target")
+            .arg(opts.target)
             .spawn()?;
         debug!("Spawned command pid={:?}", child.id());
 

